src/reduction.jl

```diff
@@ -0,0 +1,174 @@
+    rank(@view C[:,species]) == length(species)  
+end
+
+""" 
+    getinvmap(C::AbstractMatrix, elided_species::AbstractVector)
+
+Return the inverse transformation from observed species and conserved quantities
+to elided species, if it exists (checked via [`canelide`](@ref)), as a pair of matrices
+(M_kept, M_cons). The elided species are recovered via 
+`state_elided = M_kept * state_kept + M_cons * cons`.
+"""
+function getinvmap(C::AbstractMatrix, elided_species::AbstractVector)
+    @assert canelide(C, elided_species)
+    
+    n_elided = length(elided_species)
+    n_kept = size(C, 2) - n_elided
+    n_cons = size(C, 1)
+    kept_species = filter(x -> !(x in elided_species), 1:size(C, 2))
+    M_cons_float = pinv(C[:,elided_species])
+    M_cons = round.(Int, M_cons_float)
+    
+    @assert M_cons ≈ M_cons_float
+    
+    M_kept = -M_cons * view(C, :, kept_species)
+    (M_kept, M_cons)
+end
+
+"""
+    conssyms(rn::ReactionSystem, C::AbstractMatrix)
+
+Return a list of new SymbolicUtils symbols for the conserved valued determined by the
+conservation law matrix C. These become parameters for the reduced reaction system.
+
+TODO: Use gensymb to ensure uniqueness, or an alternative.
+"""
+conssyms(rn::ReactionSystem, C::AbstractMatrix) = map(id -> Sym{Number}(id), Symbol("_C", i) for i in 1:size(C,1))
+
+"""
+    getsubsdict(rn::ReactionSystem, C::AbstractMatrix, elided_species::AbstractVector, syms_cons))
+
+Return a dictionary of the form `elided_species => f(state_kept, cons)` of symbolic expressions
+expressing the elided species in terms of the kept species and the conserved quantities.
+"""
+function getsubsdict(rn::ReactionSystem, C::AbstractMatrix, elided_species::AbstractVector, syms_cons)
+    M_kept, M_cons = getinvmap(C, elided_species)
+    ret = Dict{Any,Any}()
+    syms = states(rn)
+    kept_species_idcs = filter(x -> !(x in elided_species), 1:size(C, 2))
+    kept_species = map(idx -> syms[idx], kept_species_idcs)
+    for (i, species) in enumerate(elided_species)
+        sym = syms[species]
+        ret[sym] = sum(view(M_kept, i, :) .* kept_species) + sum(view(M_cons, i, :) .* syms_cons)
+    end
+    ret
+end
+
+"""
+    redspecstoich(species, stoich, elided_species)
+
+Removes the elided species from the species and stoichiometry vectors
+(can be applied to substrates/products of a reaction).
+"""
+function redspecstoich(species, stoich, elided_species)
+    idcs = findall(x -> !(x in elided_species), species)
+    (species[idcs], stoich[idcs])
+end
+
+"""
+    jumpratelaw(rx; rxvars=get_variables(rx.rate), combinatoric_ratelaw=true)
+
+Variant of `jumpratelaw` that ignores variables not in `rxvars`, for 
+partial creation of the full jump rate law. The original doesn't seem to use 
+that parameter at all.
+"""
+function newjumpratelaw(rx; rxvars=get_variables(rx.rate), combinatoric_ratelaw=true)
+    ModelingToolkit.@unpack rate, substrates, substoich, only_use_rate = rx
```

Comment:
Parameters.jl is already available in Catalyst. Just writing `@unpack` would be fine.